File: packages/guards/src/profanity-pl.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ProfanityPlOptions {
  action: 'block' | 'warn';
}

const WORDS: string[] = [
  'kurwa', 'cholera', 'pierdolę', 'dupek', 'skurwysyn',
  'gnój', 'idiota', 'debil', 'gówno', 'zasraniec',
];

export function profanityPl(options: ProfanityPlOptions): Guard {
  return {
    name: 'profanity-pl',
    version: '0.1.0',
    description: 'Polish profanity detection',
    category: 'locale',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const matched: string[] = [];

      for (const word of WORDS) {
        const re = new RegExp(`(?:^|\\s)${word.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}(?:\\s|$)`, 'i');
        if (re.test(lower)) matched.push(word);
      }

      const triggered = matched.length > 0;

      return {
        guardName: 'profanity-pl',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Polish profanity detected' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/profanity_pl.py
"""Polish profanity detection guard."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_WORDS = [
    "kurwa", "cholera", "pierdolę", "dupek",
    "skurwysyn", "gnój", "idiota", "debil",
    "gówno", "zasraniec",
]

_PATTERNS = [
    re.compile(
        r"(?:^|\s)" + re.escape(w) + r"(?:\s|$)",
        re.IGNORECASE,
    )
    for w in _WORDS
]


class _ProfanityPl:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "profanity-pl"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []
        for i, p in enumerate(_PATTERNS):
            if p.search(text):
                matched.append(_WORDS[i])
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="profanity-pl",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                f"Polish profanity detected: {len(matched)}"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched": matched}
                if triggered
                else None
            ),
        )


def profanity_pl(
    *, action: str = "block"
) -> _ProfanityPl:
    return _ProfanityPl(action=action)
